pallets/fiat_filter/src/tests.rs

```diff
@@ -0,0 +1,804 @@
+            assert_noop!(
+                executed,
+                DispatchError::Module {
+                    index: 0,
+                    error: 3,
+                    message: Some("InsufficientBalance")
+                },
+            );
+        });
+    }
+
+    #[test]
+    fn anchor_new__Err_unsigned() {
+        ext().execute_with(|| {
+            // prepare data
+            let dat = (0..32).map(|_| rand::random()).collect();
+
+            // execute call
+            let balance_pre = <TestRt as Config>::Currency::free_balance(ALICE);
+            let call = TestCall::AnchorMod(anchor::Call::<TestRt>::deploy(dat));
+            let executed =
+                FiatFilterModule::execute_call(RawOrigin::None.into(), Box::new(call.clone()));
+            let balance_post = <TestRt as Config>::Currency::free_balance(ALICE);
+            let fee_microdock = balance_pre - balance_post;
+
+            assert_noop!(executed, DispatchError::BadOrigin);
+
+            // the call signature isn't valid, we can't charge the account any fees
+            assert_eq!(fee_microdock, 0);
+        });
+    }
+
+    #[test]
+    fn balance_transfer__Err_unexpectedCall() {
+        ext().execute_with(|| {
+            // prepare data
+            let call = TestCall::BalanceMod(pallet_balances::Call::<TestRt>::transfer(BOB, 200));
+
+            // execute call
+            let (fee_microdock, executed) = measure_fees(call);
+
+            assert_noop!(executed, Error::<TestRt>::UnexpectedCall);
+            let pdi = executed.unwrap_err();
+            assert_eq!(pdi.post_info.pays_fee, Pays::Yes);
+
+            // the call signature isn't valid, we can't charge the account any fees
+            assert_eq!(fee_microdock, 0);
+        });
+    }
+}
+
+mod tests_dock_fiat_rate {
+    use crate::*;
+    use frame_support::assert_ok;
+
+    #[test]
+    fn call_anchor_deploy__OK_different_rate() {
+        super::ext_price2().execute_with(|| {
+            use super::testrt_price2::TestRt as TestRtPrice2;
+            use super::testrt_price2::*;
+            use super::{ALICE, RATE_DOCK_USD_2};
+
+            let dat = (0..32).map(|_| rand::random()).collect();
+            let call = TestCall::AnchorMod(anchor::Call::<TestRt>::deploy(dat));
+
+            let expected_fees: u64 = 32 * PRICE_ANCHOR_OP_PER_BYTE / RATE_DOCK_USD_2 as u64;
+
+            let balance_pre = <TestRtPrice2 as Config>::Currency::free_balance(ALICE);
+            let executed =
+                FiatFilterModule::execute_call(Origin::signed(ALICE), Box::new(call.clone()));
+            let balance_post = <TestRtPrice2 as Config>::Currency::free_balance(ALICE);
+            let fee_microdock = balance_pre - balance_post;
+
+            assert_ok!(executed);
+            let pdi = executed.unwrap();
+            assert!(pdi.pays_fee == Pays::No);
+            assert_eq!(fee_microdock, expected_fees);
+        });
+    }
+}
+
+mod testrt_price2 {
```

Comment:
This is a lot of duplication for just one test with a different price.
I tried a few different ways to avoid this, none ended up as simple as duplication.

Let me know if you want me to do things differently.

Reply:
You can make struct `TestPriceProvider` have a field rather than being empty and implement a setter for that field. Then in the implementation of `PriceProvider`, return that field. Now you can use the setter in tests to change it.

Reply:
That was the first thing I tried, but if I'm not mistaken the `PriceProvider` can't access any instance field ? only type constant and associated functions ( since the methods in `trait PriceProvider` don't have `&self` as a parameter)
Or did I miss your point ?

Reply:
I tried another few ways with two PriceProvider types, a generic one with two price constant ones, but a generic `TestRuntime<_>` seemed problematic in the substrate macros I use.

Reply:
Ok, then maybe what i said wouldn't exactly work. I will try this.